npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+module Dependabot
+  module NpmAndYarn
+    class PackageName
+      DEFINITELY_TYPED_SCOPE = /types/i
+      PACKAGE_NAME_REGEX     = %r{
+          \A                                         # beginning of string
+          (?=.{1,214}\z)                             # enforce length (1 - 214)
+          (@(?<scope>[a-z0-9\-~][a-z0-9\-\._~]*)\/)? # capture 'scope' if present
+          (?<name>[a-z0-9\-~][a-z0-9\-._~]*)         # capture package name
+          \z                                         # end of string
+      }xi.freeze                                     # multi-line/case-insensitive
+
+      class InvalidPackageName < StandardError; end
+
+      def initialize(string)
+        match = PACKAGE_NAME_REGEX.match(string.to_s)
+        raise InvalidPackageName unless match
+
+        @scope = match[:scope]
+        @name  = match[:name]
+      end
+
+      def to_s
+        if scoped?
+          "@#{@scope}/#{@name}"
+        else
+          @name.to_s
+        end
+      end
+
+      def <=>(other)
+        to_s <=> other.to_s
+      end
+
+      def types_package
+        return self if types_package?
+
+        if scoped?
+          "@types/#{@scope}__#{@name}"
+        else
+          "@types/#{@name}"
+        end
+      end
```

Comment:
Sorry, I misspoke when I gave feedback about this earlier. What I meant to suggest was that this method return class instances instead of string:

```suggestion
      def types_package_name
        return nil if types_package?

        if scoped?
          PackageName.new("@types/#{@scope}__#{@name}")
        else
          PackageName.new("@types/#{@name}")
        end
      end
```

A few possible advantages:

- You can do a direct equality check to find the types package for a package (e.g. `devDependencies.include?(package.types_package_name)`
- Running through initializer again does second round of validation, which is unlikely, but possible (e.g. `@a/really_long_package_name(...up to 214 limit...)` would produce an invalid package name, because the resulting name is too long).

Reply:
...but that may be not worth the effort. In which case `return self` should be `return nil`.

```suggestion
      def types_package
        return nil if types_package?

        if scoped?
          "@types/#{@scope}__#{@name}"
        else
          "@types/#{@name}"
        end
      end
```

Reply:
Addressed a combination of these two comments in [997cade](https://github.com/dependabot/dependabot-core/pull/5001/commits/997cadecd2b823f4b4c8ece5bae0a7b7c4777c9d) 😄